internal/controllers/core/tiltfile/reconciler.go

```diff
@@ -184,6 +203,54 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	return ctrl.Result{}, nil
 }
 
+type tiltfileContentsResult struct {
+	sha256   string
+	contents []byte
+}
+
+func (r *Reconciler) reconcileTiltfileContents(ctx context.Context, tf *v1alpha1.Tiltfile, tlr *tiltfile.TiltfileLoadResult) (tiltfileContentsResult, error) {
+	result := tiltfileContentsResult{sha256: tf.Status.ContentsSHA256}
+	bytes, err := os.ReadFile(tf.Spec.Path)
+	if err != nil {
+		// If the file doesn't exist, don't do any reconciliation of contents
+		// and avoid emitting a reconciler error to the UI
+		if os.IsNotExist(err) {
+			return result, nil
+		}
+
+		return result, err
+	}
+
+	fileSha256 := fmt.Sprintf("%x", sha256.Sum256(bytes))
+	contentsSha256 := fmt.Sprintf("%x", sha256.Sum256([]byte(tf.Spec.Contents)))
+	result.sha256 = fileSha256
+
+	// File different from contents; file wins by default
+	if fileSha256 != contentsSha256 {
+		result.contents = bytes
+	}
+
+	// Short-circuit here when feature is disabled to prevent modifying disk file
+	if (tlr != nil && !tlr.FeatureFlags[feature.TiltfileEditAPI]) ||
+		(tlr == nil && !r.defaults[feature.TiltfileEditAPI].Enabled) {
+		return result, nil
+	}
+
+	// File contents changed since last hash
+	if fileSha256 != tf.Status.ContentsSHA256 {
```

Comment:
I still think the "prior" state should be read off the internal `run` object rather than the previous status - the apiserver reconciler guidelines heavily discourage reading from your own object status except in a couple of very niche cases.

Specifically, you could have a scenario where you get a cached version of the TF object so the status is outdated, and now some comparisons are happening based on "real world state" (good/`TiltfileResult`) while some are happening off potentially stale data (bad/`tf.Status`).